File: pact-jvm-server/src/main/scala/au/com/dius/pact/server/RequestRouter.scala
package au.com.dius.pact.server

import java.util
import au.com.dius.pact.core.model.{Request, Response, _}
import com.typesafe.scalalogging.StrictLogging

import scala.collection.JavaConverters._

object RequestRouter extends StrictLogging {
  def matchPath(request: Request, oldState: ServerState): Option[StatefulMockProvider] =
    (for {
      k <- oldState.getState.asScala.keys if request.getPath.startsWith(k)
      pact <- oldState.getState.asScala.get(k)
    } yield pact).headOption

  def handlePactRequest(request: Request, oldState: ServerState): Option[IResponse] =
    for {
      pact <- matchPath(request, oldState)
    } yield pact.handleRequest(request)

  def state404(request: Request, oldState: ServerState): String =
    (oldState.getState.asScala + ("path" -> request.getPath)).mkString(",\n")

  val EMPTY_MAP: util.Map[String, util.List[String]] = Map[String, util.List[String]]().asJava

  def pactDispatch(request: Request, oldState: ServerState): IResponse =
    handlePactRequest(request, oldState) getOrElse new Response(404, EMPTY_MAP,
      OptionalBody.body(state404(request, oldState).getBytes))

  def dispatch(request: Request, oldState: ServerState, config: Config): Result = {
    val urlPattern ="/(\\w*)\\?{0,1}.*".r
    val urlPattern(action) = request.getPath
    action match {
      case "create" => Create(request, oldState, config)
      case "complete" => Complete(request, oldState)
      case "publish" => Publish(request, oldState, config)
      case "" => ListServers(oldState)
      case _ => new Result(pactDispatch(request, oldState), oldState)
    }
  }
}

File: pact-jvm-server/src/main/kotlin/au/com/dius/pact/server/RequestRouter.kt
package au.com.dius.pact.server

import au.com.dius.pact.core.model.IResponse
import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.model.Request
import au.com.dius.pact.core.model.Response
import io.github.oshai.kotlinlogging.KotlinLogging

private val logger = KotlinLogging.logger {}

object RequestRouter {
  fun matchPath(request: Request, oldState: ServerState): StatefulMockProvider? {
    return oldState.state.entries.firstOrNull { request.path.startsWith(it.key) }?.value
  }

  fun handlePactRequest(request: Request, oldState: ServerState): IResponse? {
    val pact = matchPath(request, oldState)
    return pact?.handleRequest(request)
  }

  fun state404(request: Request, oldState: ServerState) =
    (oldState.state.entries.map { it.toPair() } + ("path" to request.path)).joinToString(",\n") { "${it.first} -> ${it.second}" }

  fun pactDispatch(request: Request, oldState: ServerState) =
    handlePactRequest(request, oldState) ?: Response(404, mutableMapOf(),
      OptionalBody.body(state404(request, oldState).toByteArray()))

  private val urlPattern = Regex("/(\\w*)\\?{0,1}.*")

  @JvmStatic
  fun dispatch(request: Request, oldState: ServerState, config: Config): Result {
    val matchResult = urlPattern.find(request.path)
    val (action) = matchResult!!.destructured
    return when (action) {
      "create" -> Create.apply(request, oldState, config)
      "complete" -> Complete.apply(request, oldState)
      "publish" -> Publish.apply(request, oldState, config)
      "" -> ListServers.apply(oldState)
      else -> Result(pactDispatch(request, oldState), oldState)
    }
  }
}
